app/controllers/application_controller.rb

```diff
@@ -221,9 +221,34 @@ def confirm_two_factor_authenticated(id = nil)
                                   user_needs_sp_auth_method_setup?
     return prompt_to_verify_sp_required_mfa if service_provider_mfa_policy.
                                                user_needs_sp_auth_method_verification?
+    enforce_total_session_duration_timeout
     true
   end
 
+  def enforce_total_session_duration_timeout
+    return total_session_duration_timeout if session_total_duration_expired?
+    ensure_user_session_has_created_at
+  end
+
+  def total_session_duration_timeout
+    sign_out
+    flash[:info] = t('devise.failure.timeout')
+    redirect_to root_url
+  end
```

Comment:
Can we rename this method? Based on the name it would just return a number and not have any side effects?

How about `sign_out_with_timeout_error`?